Mathlib/Topology/UniformSpace/Ultra.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2017 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Topology.UniformSpace.Basic
+import Mathlib.Topology.Bases
+
+/-!
+# Ultrametric (nonarchimedean) uniform spaces
+
+Ultrametric (nonarchimedean) uniform spaces are ones that generalize ultrametric spaces by
+having a uniformity based on equivalence relations.
+
+## Main definitions
+
+In this file we define `IsUltraUniformity`, a Prop mixin typeclass.
+
+## Main results
+
+* `TopologicalSpace.isTopologicalBasis_clopens`: a uniform space with a nonarchimedean uniformity
+  has a topological basis of clopen sets in the topology, meaning that it is topologically
+  zero-dimensional.
+
+## Implementation notes
+
+As in the `Topology/UniformSpace/Defs.lean` file, we do not reuse `Data/Rel.lean`
+but rather extend the relation properties as needed.
+
+## TODOs
+
+* Prove that `IsUltraUniformity` iff `IsUltrametricDist` on a `PseudoMetricSpace`
+* Generalize `IsUltrametricDist` to `IsUltrametricUniformity`
+* Provide `IsUltraUniformity` for the uniformity in a `Valued` ring
+* Generalize results about open/closed balls and spheres in `IsUltraUniformity` to
+  combine applications for `MetricSpace.ball` and valued "balls"
+* Use `IsUltraUniformity` to work with profinite/totally separated spaces
+* Define the nonarchimedean uniformity of a space that is a product of `IsUltraUniformity`s
+* Show that the `UniformSpace.Completion` of an `IsUltraUniformity` is `IsUltraUniformity`
+
+## References
+
+* [D. Windisch, *Equivalent characterizations of non-Archimedean uniform spaces*][windisch2021]
+* [A. C. M. van Rooij, *Non-Archimedean uniformities*][vanrooij1970]
+
+-/
+
+open Set Filter Topology
+open scoped Uniformity
+
+variable {X : Type*}
+
+/-- The relation is transitive. -/
+def TransitiveRel (V : Set (X × X)) : Prop :=
+  ∀ ⦃x y z⦄, (x, y) ∈ V → (y, z) ∈ V → (x, z) ∈ V
+
+lemma TransitiveRel.comp_eq_of_idRel_subset {s : Set (X × X)}
+    (h : TransitiveRel s) (h' : idRel ⊆ s) :
+    s ○ s = s := by
+  refine le_antisymm ?_ (subset_comp_self h')
+  intro ⟨x, y⟩
+  simp only [mem_compRel, forall_exists_index, and_imp]
+  intro z
+  exact @h x z y
```

Comment:
Should this stay here? Or move to some basic file on sets?
Also, should the name be `IsTransitiveRel`?

Reply:
I followed the style in `UniformSpace/Defs.lean` which has the following explanation and defs
```
There is already a theory of relations in `Data/Rel.lean` where the main definition is
`def Rel (α β : Type*) := α → β → Prop`.
The relations used in the current file involve only one type, but this is not the reason why
we don't reuse `Data/Rel.lean`. We use `Set (α × α)`
instead of `Rel α α` because we really need sets to use the filter library, and elements
of filters on `α × α` have type `Set (α × α)`.
[...]
/-- The relation is invariant under swapping factors. -/
def SymmetricRel (V : Set (α × α)) : Prop :=
  Prod.swap ⁻¹' V = V
```

Reply:
How about #22594?